src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs

```diff
@@ -3,217 +3,65 @@
-            {
-                var tcs2 = (TaskCompletionSource<int>)state2;
-                if (task2.IsCanceled)
-                {
-                    tcs2.SetCanceled();
-                }
-                else if (task2.IsFaulted)
-                {
-                    tcs2.SetException(task2.Exception);
-                }
-                else
-                {
-                    tcs2.SetResult(task2.Result);
-                }
-            }, tcs, cancellationToken);
-            return tcs.Task;
+            _pipeReader = pipeReader;
         }
 
         public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
-            ValidateState(cancellationToken);
-
             return ReadAsyncInternal(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
         }
 
         public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
         {
-            ValidateState(cancellationToken);
-
             return ReadAsyncInternal(destination, cancellationToken);
         }
 
-        private async ValueTask<int> ReadAsyncInternal(Memory<byte> buffer, CancellationToken cancellationToken)
-        {
-            try
-            {
-                return await _body.ReadAsync(buffer, cancellationToken);
-            }
-            catch (ConnectionAbortedException ex)
-            {
-                throw new TaskCanceledException("The request was aborted", ex);
-            }
-        }
-
-        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        public override int Read(byte[] buffer, int offset, int count)
         {
-            if (destination == null)
-            {
-                throw new ArgumentNullException(nameof(destination));
-            }
-            if (bufferSize <= 0)
+            if (!_bodyControl.AllowSynchronousIO)
             {
-                throw new ArgumentException(CoreStrings.PositiveNumberRequired, nameof(bufferSize));
+                throw new InvalidOperationException(CoreStrings.SynchronousReadsDisallowed);
             }
 
-            ValidateState(cancellationToken);
-
-            return CopyToAsyncInternal(destination, cancellationToken);
+            return ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
         }
 
-        private async Task CopyToAsyncInternal(Stream destination, CancellationToken cancellationToken)
+        private ValueTask<int> ReadAsyncInternal(Memory<byte> buffer, CancellationToken cancellationToken)
         {
             try
             {
-                await _body.CopyToAsync(destination, cancellationToken);
+                return base.ReadAsync(buffer, cancellationToken);
             }
             catch (ConnectionAbortedException ex)
             {
                 throw new TaskCanceledException("The request was aborted", ex);
```

Comment:
I had to change a few tests because this exception was not being caught. Definitely need to fix. I'm fairly sure it's due to new async state machines being introduced.

Reply:
It's the await removal that does this. Both ConnectionAbortedException and TaskCanceledExceptions are OperationCanceledExceptions, so if we have to break, there's a workaround, but let's not break.